### PR TITLE
New Swap Mode

### LINKED ISSUE #19: Support replacement
Replace public registry to private one by custom configuration

### LINKED ISSUE #27: Imageswap swaps out complete registry and repository path
**What happened**:

Deploy an image with definition `somepublic.io/repo/subrepo/image:4.0`.
Using `IMAGE_PREFIX=someprivate.io` to swap the registry, the resulting image was `someprivate.io/image:4.0` (note the missing paths).

**What you expected to happen**:

I would have expected the resulting image to be `someprivate.io/repo/subrepo/image:4.0`.

**How to reproduce it (as minimally and precisely as possible)**:

Take any kubernetes resource with an image with more than one slash in the path and look at the replacement.

**Anything else we need to know?**:

If this is indeed intended behaviour, I think it would be valuable to control it somehwo to allow partial replacement to keep repository structure. This is interesting when having multiple internal registries (e.g. dev + prod).

**Environment**:

- Kubernetes version (use `kubectl version`): 1.18.17
- Cloud provider or hardware configuration: AKS (Azure)
- Others: tried with latest release + master of imageswap-webhook

### PR DESCRIPTION
**Pull request by phenixblue**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation

/kind feature

**What this PR does / why we need it**:

This PR implements a new "MAPS" mode for the image swapping logic.

>MAPS Mode enables a high degree of flexibility for the ImageSwap Webhook.
>
>In MAPS mode, the webhook reads from a `map file` that defines one or more mappings (key/value pairs) for imageswap >logic. With the `map file` configuration, swap logic for multiple registries and patterns can be configured. In contrast, the >LEGACY mode only allowed for a single `IMAGE_PREFIX` to be defined for image swaps.

**Which issue(s) this PR fixes**:

Fixes #19 
Fixes #27 

**Special notes for your reviewer**:

LEGACY mode is maintained in this release, but is not the default. Coments have been added to note for future cleanup of LEGACY mode. While Python unit-tests continue to test LEGACY mode functionality, the end-to-end/functional testing does not.

**Does this PR introduce a user-facing change?**:

```release-note

ImageSwap v1.4.0 has major changes

There is a new "MAPS" mode logic that has been added to allow for more flexibility in the image swapping logic.

The existing logic, referred to as "LEGACY" mode, is still available, but has been deprecated.

To continue using the "LEGACY" mode logic set the "IMAGESWAP_MODE" environment variable accordingly.

Image Definition Preservation: Updates have been made to how image definitions are processed during a swap. Previously the swap logic would drop the image org/project before adding the prefix (ie. `nginx/nginx-ingress:latest` would drop the `nginx/` portion of the image definition).

In v1.4.0+ the swap logic will preserve all parts of the image except the Registry (ie. `docker.io/nginx/nginx-ingress` will drop the `docker.io` only from the image definition).
```

**Additional documentation e.g., usage docs, etc.**:

```docs

```

### CODE REVIEW

File: app/imageswap/test/test_image_patterns.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+        expected_image = "my.example.com/mirror-docker.io/nginx:latest"
+        result = imageswap.swap_image(container_spec)
+
+        self.assertTrue(result)
+        self.assertEqual(container_spec["image"], expected_image)
+
+    def test_image_format_project_image(self):
+
+        """Method to test MAP based swap (project+image: \"nginx:latest\")"""
```
Comment (ilrudie) [inline app/imageswap/test/test_image_patterns.py:72]: **Suggestion:** Looks like maybe a typo in this output. Says "nginx:latest" but then test used "ubuntu/ubuntu".

**Review by ilrudie — CHANGES_REQUESTED** [review]
Just one small suggestion. Otherwise looks good.

File: app/imageswap/test/test_image_patterns.py (hunk 1 above)
Comment (phenixblue) [reply app/imageswap/test/test_image_patterns.py]: Fixed

Comment (ilrudie) [reply app/imageswap/test/test_image_patterns.py]: Thanks!

**Review by ilrudie — APPROVED** [review]

**Comment by ilrudie** [comment]
LGTM

**Pull request merged**